src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs

```diff
@@ -62,25 +62,51 @@ public override TypeSyntax VisitArrayType(IArrayTypeSymbol symbol)
             {
                 ThrowIfNameOnly();
 
-                var underlyingNonArrayType = symbol.ElementType;
-                while (underlyingNonArrayType.Kind == SymbolKind.ArrayType)
+                ITypeSymbol underlyingType = symbol;
+
+                while (underlyingType is IArrayTypeSymbol innerArray)
                 {
-                    underlyingNonArrayType = ((IArrayTypeSymbol)underlyingNonArrayType).ElementType;
+                    underlyingType = innerArray.GetElementTypeWithAnnotatedNullability();
+
+                    if (underlyingType.GetNullability() == NullableAnnotation.Annotated)
+                    {
+                        // If the inner array we just moved to is also nullable, then
+                        // we must terminate the digging now so we produce the syntax for that,
+                        // and then append the ranks we passed through at the end. This is because
+                        // nullability annotations acts as a "barrier" where we won't reorder array
+                        // through. So whereas:
+                        //
+                        //     string[][,]
+                        //
+                        // is really an array of rank 1 that has an element of rank 2,
+                        //
+                        //     string[]?[,]
+                        //
+                        // is really an array of rank 2 that has nullable elements of rank 1.
+
+                        break;
+                    }
                 }
 
-                var elementTypeSyntax = underlyingNonArrayType.GenerateTypeSyntax();
+                var elementTypeSyntax = underlyingType.GenerateTypeSyntax();
                 var ranks = new List<ArrayRankSpecifierSyntax>();
 
                 var arrayType = symbol;
-                while (arrayType != null)
+                while (arrayType != null && !arrayType.Equals(underlyingType))
```

Comment:
any concern about the right nullability comparison happening here?

Reply:
It'll be doing the non-nullable-aware compare but the "interesting" part of the comparison anyways is the actual array depth. Switching to nullable aware would be the same behavior.

Reply:
SGTM!